ethkey/cli/src/main.rs

```diff
@@ -214,12 +249,65 @@ fn execute<S, I>(command: I) -> Result<String, Error> where I: IntoIterator<Item
 			let address = args.arg_address.parse().map_err(|_| EthkeyError::InvalidAddress)?;
 			verify_address(&address, &signature, &message)?
 		} else {
-			unreachable!();
+			return Ok(format!("{}", USAGE))
 		};
 		Ok(format!("{}", ok))
+	} else if args.cmd_recover {
+		let display_mode = DisplayMode::new(&args);
+		let known_phrase = args.arg_known_phrase;
+		let address = args.arg_address.parse().map_err(|_| EthkeyError::InvalidAddress)?;
+		let phrase = in_threads(move |idx, max| {
+			let known_phrase = known_phrase.clone();
+			move || brain_recover(&address, &known_phrase, BRAIN_WORDS, Some((idx, max)))
+		})?;
+		let keypair = Brain::new(phrase.clone()).generate().expect("Brain wallet generator is infallible; qed");
+		Ok(display((keypair, Some(phrase)), display_mode))
 	} else {
-		unreachable!();
+		Ok(format!("{}", USAGE))
+	}
+}
+
+const BRAIN_WORDS: usize = 12;
+
+fn validate_phrase(phrase: &str) -> String {
+	match Brain::validate_phrase(phrase, BRAIN_WORDS) {
+		Ok(()) => format!("The recovery phrase looks correct.\n"),
+		Err(err) => format!("The recover phrase was not generated by Parity: {}", err)
+	}
+}
+
+fn in_threads<F, X, O>(task: F) -> Result<O, EthkeyError> where
+	O: Send + 'static,
+	X: Send + 'static,
+	F: Fn(usize, usize) -> X,
+	X: FnOnce() -> Result<O, EthkeyError>,
+{
+	let mut handles = Vec::new();
+	let max = num_cpus::get();
+
+	let (tx, rx) = sync::mpsc::sync_channel(1);
+	for i in 0..max {
+		let t = task(i, max);
+		let tx = tx.clone();
+		handles.push(thread::spawn(move || match t() {
+			Ok(solution) => {
+				tx.send(solution).expect("Receiving end never dropped.");
+				Ok(())
+			},
+			Err(err) => Err(err),
+		}));
 	}
+
+	if let Ok(solution) = rx.recv() {
```

Comment:
Is it possible that `recv` will return `Err` here? It only fails when sender is destroyed an here are `max+1` senders, of which this last 1 is never destroyed. So it looks like that if all threads have failed with an error, we will hang in this `recv` call. I can be missing something, though :)

Reply:
Good points, thanks. Will addresss both by switching to `rayon` as suggested by @0x7CFE